File: Client/src/main/java/com/rspsi/editor/model/BridgeLink.java
package com.rspsi.editor.model;

import java.util.Objects;

/** Explicit authored-plane to effective-plane relationship for a bridge tile. */
public record BridgeLink(TileCoordinate upper, TileCoordinate lower) {
    public BridgeLink {
        upper = Objects.requireNonNull(upper, "upper");
        lower = Objects.requireNonNull(lower, "lower");
        if (upper.x() != lower.x() || upper.y() != lower.y()
                || upper.plane() != lower.plane() + 1) {
            throw new IllegalArgumentException("Bridge links must connect adjacent planes at one tile");
        }
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/BridgeLink.kt
package com.rspsi.editor.model

/**
 * Explicit authored-plane to effective-plane relationship for one bridge tile.
 *
 * Bridge links are positional metadata, not renderer state: upper and lower must refer to
 * the same X/Y coordinate on adjacent planes. Kept as a JVM record for Java compatibility.
 */
@JvmRecord
data class BridgeLink(
    val upper: TileCoordinate,
    val lower: TileCoordinate,
) {
    init {
        if (
            upper.x != lower.x ||
            upper.y != lower.y ||
            upper.plane != lower.plane + 1
        ) {
            throw IllegalArgumentException(
                "Bridge links must connect adjacent planes at one tile",
            )
        }
    }

    override fun toString(): String =
        "BridgeLink[upper=$upper, lower=$lower]"
}
